### PR TITLE
Add Chunked SimPO Loss

### PR DESCRIPTION
**Pull request by pramodith**
## Summary
This PR adds the Simple Preference Optimization Loss function. The only difference between SimPO and CPO is a margin term `gamma` which specifies that the preferred response should be atleast gamma logit points better than the losing response.

$$SimPOLoss = -\log(\sigma(\beta\log(\pi_\theta(y_c|x)) - \beta\log(\pi_\theta(y_r|x)) - \gamma))$$

Note that SimPO explicitly specifies that $$\pi_\theta(y|x)$$ needs to be normalized by length, unlike DPO.

This corresponds to Eq 6 in the [paper](https://arxiv.org/pdf/2405.14734).

## Testing Done
GPU A100-80G-SXM

Screenshot 2024-11-15 at 2 38 23 PM
Screenshot 2024-11-15 at 2 38 37 PM

- Hardware Type: <BLANK>
- [X] run `make test` to ensure correctness
- [X] run `make checkstyle` to ensure code style
- [X] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by shivam15s** [comment]
Great PR @pramodith ! Could you rebase?

**Comment by pramodith** [comment]
> Great PR @pramodith ! Could you rebase?

Good to go now!

**Review by shivam15s — APPROVED** [review]
lgtm
